src/main/scala/mesosphere/marathon/upgrade/GroupVersioningUtil.scala

```diff
@@ -59,4 +60,53 @@ object GroupVersioningUtil {
     }
   }
 
+  /**
+    * Calculate a new group from the given `to` parameter that sets the version of all changed pods
+    * to the given `version`.
+    *
+    * The `RootGroup` object returned from this function, has its version set to the given `version`.
+    * Therefore even if there are no changed pods, the version of `RootGroup` object returned is always
+    * set to `version`.
+    *
+    * @param version the version of all changed pods
+    * @param from the original group
+    * @param to the updated group
+    * @return the updated group with updated pof versions
+    */
+  def updateVersionInfoForChangedPods(version: Timestamp, from: RootGroup, to: RootGroup): RootGroup = {
+
+    def updatePodVersionInfo(maybeOldPod: Option[PodDefinition], newPod: PodDefinition): PodDefinition = {
+      val newVersionInfo = maybeOldPod match {
+        case None =>
+          log.info(s"${newPod.id}: new pod detected")
+          VersionInfo.forNewConfig(newVersion = version)
+        case Some(oldPod) =>
+          if (oldPod.isUpgrade(newPod)) {
+            log.info(s"${newPod.id}: upgrade detected for Pod (oldVersion ${oldPod.versionInfo})")
+            oldPod.versionInfo.withConfigChange(newVersion = version)
+          } else if (oldPod.isOnlyScaleChange(newPod)) {
+            log.info(s"${newPod.id}: scaling op detected for Pod (oldVersion ${oldPod.versionInfo})")
+            oldPod.versionInfo.withScaleOrRestartChange(newVersion = version)
+          } else if (oldPod.versionInfo != newPod.versionInfo && newPod.versionInfo == VersionInfo.NoVersion) {
+            log.info(s"${newPod.id}: restart detected for Pod (oldVersion ${oldPod.versionInfo})")
+            oldPod.versionInfo.withScaleOrRestartChange(newVersion = version)
+          } else {
+            oldPod.versionInfo
```

Comment:
unfortunate that we still execute the copy a few lines later, even if this doesn't change. only to later throw out the object later on when the versions are detected as unchanged

Reply:
It's unfortunate that we have this logic twice - for apps and pods. What do you think about moving `versionInfo` into `RunSpec`?

Reply:
I kind of hate `versionInfo` and I also despise code duplication. If `versionInfo` is refactored into RunSpec then maybe tackle that in a separate PR? I'd love to see `versionInfo` disappear (and be replaced by a cleaner data structure) because it conflates "object version" (static metadata) with status-related things (the last time this was scaled, or some other config change was made).